### PR TITLE
Fix top-k and top-p sampling in CausalLM Application

### PR DESCRIPTION
**Pull request by haehun**
## Dependency of the PR
None

## Commits to be reviewed in this PR

<details><summary>Fix top-k and top-p sampling in CausalLM Application</summary><br />

As @jaemini-shin pointed out in #3709, `applyTKP` function in CausalLM application did not apply top-k and top-p sampling at all.

This PR correctly applies them, using same code with LLaMA application.

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Haehun Yang <haehun.yang@samsung.com>

</details>

### Summary

- Fix top-k and top-p sampling in CausalLM Application

Signed-off-by: Haehun Yang <haehun.yang@samsung.com>

### CODE REVIEW

File: Applications/CausalLM/llm_util.cpp [hunk 1]
```diff
@@ -82,9 +82,23 @@ float applyTKP(float *logits, int len, float temperature, unsigned int top_k,
...
+    ++top_index;
+  }
+
+  // Apply Top-K and Top-P
+  std::fill_n(logits, len, -INFINITY);
+  for (unsigned int i = 0; i < top_index; ++i) {
+    logits[top_indices_and_logits[i].first] = top_indices_and_logits[i].second;
+  }
+
   return top_indices_and_logits[0].second;
```
Comment (jaemini-shin) [inline Applications/CausalLM/llm_util.cpp:103]: This logic is the same as existing one!

https://github.com/nntrainer/nntrainer/blob/f22525c1e8debb8f39b333623495a321c3c8bb74/Applications/LLaMA/jni/main.cpp#L85-L104

**Review by jaemini-shin — APPROVED** [review]
LGTM!

File: Applications/CausalLM/llm_util.cpp [hunk 2]
```diff
@@ -82,9 +82,23 @@ float applyTKP(float *logits, int len, float temperature, unsigned int top_k,
...
                     top_indices_and_logits.end(),
                     [](auto &a, auto &b) { return a.second > b.second; });
 
+  // Accumulate logits
+  float cum_prob = 0;
+  unsigned int top_index = 0;
+  while (cum_prob <= top_p && top_index < top_k) {
+    cum_prob += top_indices_and_logits[top_index].second;
+    ++top_index;
+  }
```
Comment (songgot) [inline Applications/CausalLM/llm_util.cpp:95]: need to apply Softmax on top_indices_and_logits to convert raw logits into probabilities before calculating the cumulative probability. Right?
```
  float max_logit = top_indices_and_logits[0].second;
  double sum_exp = 0.0;
  // Calculate sum of exponentials for the top-k elements
  for (unsigned int i = 0; i < top_k; ++i) {
    sum_exp += std::exp(top_indices_and_logits[i].second - max_logit);
  }
  // Accumulate probabilities for Top-P (Nucleus Sampling)
  float cum_prob = 0.0f;
  unsigned int top_index = 0;

  while (top_index < top_k) {
    // Convert current logit to probability
    float prob = std::exp(top_indices_and_logits[top_index].second - max_logit) / sum_exp;
    cum_prob += prob;
    
    // Increment index first to include at least one token
    ++top_index;
    
    if (cum_prob >= top_p) {
      break;
    }
  }
  ```

Comment (jaemini-shin) [reply Applications/CausalLM/llm_util.cpp:95]: I think you're right, we should accumulate probabilities not logits to compare with top_p

Comment (haehun) [reply Applications/CausalLM/llm_util.cpp:95]: Oh that's right...
